### ISSUE TITLE
dask.array buggy with object dtype

### ISSUE THREAD

**Issue by dcherian** [dask/dask#11585]
**Describe the issue**:

**Minimal Complete Verifiable Example**:

```python
import dask.array
import pandas as pd
import numpy as np

idx = pd.MultiIndex.from_product([list("abc"), [0, 1]])
dask.array.from_array(idx.values, chunks=-1)[0].compute()
```

Interestingly
```
array = np.array([('a', 0), ('a', 1), ('b', 0), ('b', 1), ('c', 0), ('c', 1)], dtype=object)
dask.array.from_array(array, chunks=-1)[0].compute()
```

succeeds :/ even though that should be identical to `idx.values`

**Anything else we need to know?**:

**Environment**:

- Dask version:
- Python version:
- Operating System:
- Install method (conda, pip, source):

**Comment by phofl** [comment]
That's not exactly the same unfortunately :(

The first creates an array of tuples with one dimension, the second one an array with 2 dimensions but scalars instead of tuples

**Comment by phofl** [comment]
This is tricky, not sure how to tackle this in an elegant way yet

**Comment by dcherian** [comment]
Right, my bad on the two examples not being the same. What changed here suddenly?

**Comment by phofl** [comment]
Did this work before?

**Comment by phofl** [comment]
Oh, thanks. That helps. I'll dig into what worked differently before.

**Comment by dcherian** [comment]
It's confusing but this is the failing test: https://github.com/pydata/xarray/blob/86a408d20a946efbd14c3f1beb72bd13ebd6a481/xarray/tests/test_variable.py#L673-L677 That PR of mine xfailed two tests. this is the second one. This test is inherited by `TestVariableWithDask` (or something like that)

**Comment by phofl** [comment]
Thanks, I completely missed that.

It took me a while to understand what's actually going on here (it's a PITA). I think the new behavior is correct and the old one was buggy.

We have the following test:

```
idx = pd.MultiIndex.from_product([list("abc"), [0, 1]])
v = Variable("x", idx).chunk()
x = v[0]
assert_identical(Variable((), ("a", 0)), x)
```

Basically, we create a Variable with 6 elements and then select the first one. The data of the variable is stored in a **dask array**, not a pandas MultiIndex like the in memory version. Indexing the first element gives us back a scalar, i.e. the tuple ("a", 0) for both versions. 

Xarray has some logic to store the tuple back in an array for the pandas MultiIndex, but not for Dask arrays.


Now what changed: Dask had some weird code that was optimizing different types of indexing operations together into one. This logic had a weird bug that stored the result in an array instead of giving us back a scalar. [This PR](https://github.com/dask/dask/pull/11525) replaced the overly complicated thing with a new implementation that is much simpler but instead runs every indexing operation after the previous one, i.e. no fusion of different indexing operations. This means that we follow NumPy behavior very closely instead of doing our own thing. 

How can we validate this?

The change in behavior only shows up when we have the selection of the first element immediately after chunking the variable. If we insert a random operation in between, the test will raise on an older Dask version as well, i.e.

```
idx = pd.MultiIndex.from_product([list("abc"), [0, 1]])
v = Variable("x", idx).chunk()
x = v.fillna("bla")[0]
assert_identical(Variable((), ("a", 0)), x)
```

The fillna disables the weird optimization thing that's going on and thus we get NumPy semantics under the hood. 

TLDR: The test worked accidentally because Dask had a bug in how it treated consecutive indexing operations.

What to do? Not quite sure, I suspect that Xarray might want some logic to handle the Dask case similar to the numpy case?

**Comment by nitinsharmacs** [comment]
Hi. Any update on this issue?
